lib/evmone_precompiles/kzg.cpp

```diff
@@ -0,0 +1,155 @@
+#include "kzg.hpp"
+#include "sha256.hpp"
+#include <blst.h>
+#include <algorithm>
+#include <optional>
+#include <span>
+
+namespace evmone::crypto
+{
+namespace
+{
+/// The field element 1 in Montgomery form.
+constexpr blst_fp ONE = {0x760900000002fffd, 0xebf4000bc40c0002, 0x5f48985753c758ba,
+    0x77ce585370525745, 0x5c071a97a256ec6d, 0x15f65ec3fa80e493};
+
+/// The negation of the subgroup G1 generator -[1]₁ (Jacobian coordinates in Montgomery form).
+constexpr blst_p1 G1_GENERATOR_NEGATIVE = {
+    {0x5cb38790fd530c16, 0x7817fc679976fff5, 0x154f95c7143ba1c1, 0xf0ae6acdf3d0e747,
+        0xedce6ecc21dbf440, 0x120177419e0bfb75},
+    {0xff526c2af318883a, 0x92899ce4383b0270, 0x89d7738d9fa9d055, 0x12caf35ba344c12a,
+        0x3cff1b76964b5317, 0x0e44d2ede9774430},
+    ONE};
+
+/// The negation of the subgroup G2 generator -[1]₂ (Jacobian coordinates in Montgomery form).
+constexpr blst_p2 G2_GENERATOR_NEGATIVE{
+    {{{0xf5f28fa202940a10, 0xb3f5fb2687b4961a, 0xa1a893b53e2ae580, 0x9894999d1a3caee9,
+          0x6f67b7631863366b, 0x058191924350bcd7},
+        {0xa5a9c0759e23f606, 0xaaa0c59dbccd60c3, 0x3bb17e18e2867806, 0x1b1ab6cc8541b367,
+            0xc2b6ed0ef2158547, 0x11922a097360edf3}}},
+    {{{0x6d8bf5079fb65e61, 0xc52f05df531d63a5, 0x7f4a4d344ca692c9, 0xa887959b8577c95f,
+          0x4347fe40525c8734, 0x197d145bbaff0bb5},
+        {0x0c3e036d209afa4e, 0x0601d8f4863f9e23, 0xe0832636bacc0a84, 0xeb2def362a476f84,
+            0x64044f659f0ee1e9, 0x0ed54f48d5a1caa7}}},
+    {{ONE, {}}}};
+
+/// The point from the G2 series, index 1 of the Ethereum KZG trusted setup,
+/// i.e. [s]₂ where s is the trusted setup's secret.
+/// Affine coordinates in Montgomery form.
+constexpr blst_p2_affine KZG_SETUP_G2_1{
+    {{{0x6120a2099b0379f9, 0xa2df815cb8210e4e, 0xcb57be5577bd3d4f, 0x62da0ea89a0c93f8,
+          0x02e0ee16968e150d, 0x171f09aea833acd5},
+        {0x11a3670749dfd455, 0x04991d7b3abffadc, 0x85446a8e14437f41, 0x27174e7b4e76e3f2,
+            0x7bfa6dd397f60a20, 0x02fcc329ac07080f}}},
+    {{{0xaa130838793b2317, 0xe236dd220f891637, 0x6502782925760980, 0xd05c25f60557ec89,
+          0x6095767a44064474, 0x185693917080d405},
+        {0x549f9e175b03dc0a, 0x32c0c95a77106cfe, 0x64a74eae5705d080, 0x53deeaf56659ed9e,
+            0x09a1d368508afb93, 0x12cf3a4525b5e9bd}}}};
+
+/// Load and validate an element from the group order field.
+std::optional<blst_scalar> validate_scalar(std::span<const std::byte, 32> b) noexcept
+{
+    blst_scalar v;
+    blst_scalar_from_bendian(&v, reinterpret_cast<const uint8_t*>(b.data()));
+    return blst_scalar_fr_check(&v) ? std::optional{v} : std::nullopt;
+}
+
+/// Uncompress and validate a point from G1 subgroup.
+std::optional<blst_p1_affine> validate_G1(std::span<const std::byte, 48> b) noexcept
```

Comment:
I haven't looked at the rest of the code, so this may not be easy; the 32 and 48 are well defined constants in the spec -- is it not possible to use them instead of using magic numbers?

Reply:
Yes, make sense. I plan to work on this but I need a separate PR for this to explore some directions (e.g. use `std::span`).